src/api/unified/symbol_manager.cpp

```diff
@@ -171,12 +171,18 @@ AFSymbolManager::AFSymbolManager()
     : activeHandle(nullptr)
     , defaultHandle(nullptr)
     , numBackends(0)
+    , numBackendHandles(NUM_BACKENDS)
```

Comment:
So we could do that, but this will necessitate other changes in the code to produce an equivalent result as before. I think if we do this, I have to rename this variable in order to reflect its purpose better (and differentiate it from `numBackends`). See the proposed changes below:

```diff
modified   src/api/unified/symbol_manager.cpp
@@ -171,7 +171,7 @@ AFSymbolManager::AFSymbolManager()
     : activeHandle(nullptr)
     , defaultHandle(nullptr)
     , numBackends(0)
-    , numBackendHandles(NUM_BACKENDS)
+    , numCustomHandles(0)
     , backendsAvailable(0)
     , logger(loggerFactory("unified")) {
     // In order of priority.
@@ -239,7 +239,7 @@ af_err AFSymbolManager::setBackend(af::Backend bknd) {
 }
 
 af_err AFSymbolManager::addBackendLibrary(const char *lib_path) {
-    if ((numBackendHandles + 1) > MAX_BKND_HANDLES) {
+    if ((NUM_BACKENDS + numCustomHandles + 1) > MAX_BKND_HANDLES) {
         // No more space for an additional handle
         UNIFIED_ERROR_LOAD_LIB();
     }
@@ -266,8 +266,8 @@ af_err AFSymbolManager::addBackendLibrary(const char *lib_path) {
 
         if (show_load_path) { printf("Using %s\n", lib_path); }
 
-        bkndHandles[numBackendHandles] = handle;
-        numBackendHandles++;
+        bkndHandles[NUM_BACKENDS + numCustomHandles] = handle;
+        numCustomHandles++;
 
         return AF_SUCCESS;
     }
```

Reply:
I think the new approach is easy to understand what the earlier variable `numBackendHandles`'s purpose is.

However, I believe we can still retain old logic if we just rename `numBackendHandles` to `customBackendHandleIndex` or `customBackendTracker` or something like that.

Reply:
Thanks. However, the indexing into the `bkndHandles` array have to be adjusted if we initialize `numBackendHandles` or `numCustomHandles` to 0. Originally, the purpose of `numBackendHandles` is to count how many handles does `bkndHandles` contain. I initialized it to 3 to reserve space for the default 3 backend handles (regardless of whether the machine or the build actually supports all 3 backends), because of the way the `AFSymbolManager` constructor loads the libraries and indexes into `bkndHandles`:

```cpp
/*** af/defines.h ***/
// These enums should be 2^x
typedef enum {
    AF_BACKEND_DEFAULT = 0,  ///< (Mark: don't mind the following comment, it is something we need to correct) Default backend order: OpenCL -> CUDA -> CPU
    AF_BACKEND_CPU     = 1,  ///< CPU a.k.a sequential algorithms
    AF_BACKEND_CUDA    = 2,  ///< CUDA Compute Backend
    AF_BACKEND_OPENCL  = 4   ///< OpenCL Compute Backend
} af_backend;

/*** symbol_manager.cpp ***/
    // In order of priority.
    // (Mark: this is the actual order that is implemented, not the comment above in the header)
    static const af_backend order[] = {AF_BACKEND_CUDA, AF_BACKEND_OPENCL,
                                       AF_BACKEND_CPU};

    // Decremeting loop. The last successful backend loaded will be the most
    // prefered one.
    for (int i = NUM_BACKENDS - 1; i >= 0; i--) {
        int backend          = order[i] >> 1;  // 2 4 1 -> 1 2 0
        bkndHandles[backend] = openDynLibrary(order[i]);
        if (bkndHandles[backend]) {
            activeHandle  = bkndHandles[backend];
            activeBackend = (af_backend)order[i];
            numBackends++;
            backendsAvailable += order[i];
        }
    }
```

We can see that the index `backend` is calculated based on the enum values of the backends. Notice that `backend` is the same no matter how many backends are actually available. Let's take the worst case scenario: suppose that only OpenCL is available. `AF_BACKEND_OPENCL` is 4, and shifting it left by 1 makes it 2, and so `bkndHandles[2]` now contains the handle for the libafopencl library. That means any custom libraries that will be added must take `bkndHandles[3]` and beyond. `NUM_BACKENDS` is 3 and that's why I initialized `numBackendHandles` to `NUM_BACKENDS` (`bkndHandles` is using `numBackendHandles` as its index).